torchdata/stateful_dataloader/worker.py

```diff
@@ -111,16 +120,20 @@ def _worker_loop(
 
         from torch.utils.data import _DatasetKind
 
-        incremental_worker_state = _IncrementalWorkerState(worker_state)
+        incremental_worker_state: _IncrementalWorkerState
         init_exception = None
         fetcher = None
-
+        initial_state = None
         try:
             if init_fn is not None:
                 init_fn(worker_id)
 
             fetcher = _DatasetKind.create_fetcher(dataset_kind, dataset, auto_collation, collate_fn, drop_last)
 
+            # See NOTE [ Incremental worker state ]
+            initial_state = worker_state or _make_state_dict(worker_id, dataset_kind, fetcher, dataset)
```

Comment:
I'm keeping the flow identical for now between worker_state == None and != None to reduce the differences in code paths between the two cases, and easier to reason about that main and worker start with the same state. We can optimize this in a subsequent diff by not sending worker_state back if we just got it from the main proc, which will be a trivial change, but IMO makes it a little harder to follow